### PR TITLE
[ARCTIC-395][SPARK]Feature merge into

### LINKED ISSUE #395: [Feature][Spark]: Support Merge Into for Spark3.x
### Description

suport  merge into for spark 3.x version

### Use case/motivation

spark 3.x has already define merge into sql grammar, but implement is miss, and iceberg has already implement merge into.  

arctic use iceberg as unkeyed table implement , but keyed table , hive compatibility table doesn't support merge now. 

merge into sql is useful in some case, so we shoud implemnt this for all arctic table type under spark 3.x version.

### Describe the solution

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hellojinsilei**
## Why are the changes needed?

resolve #395 
This PR implements merge-on-read MERGE INTO in Spark 3.1.
Merge into syntax:

```
merge into <target_table> as <alias_name_t> 
using <source expression|table_name> as <alias_name_s> on <boolean expression>
when matched [and <boolean expression2>] then update set <set_clause_list>
when matched [and <boolean expression3>] then delete 
when not matched [and <boolean expression4>] then insert values <value_list>
```

## Brief change log

Add write mode: MERGE

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            val insertProjection = buildInsertProjection(source)
+            val join = insertAction.condition match {
+              case Some(insertCond) => Filter(insertCond, insertProjection)
+              case None => insertProjection
+            }
+            val joinHint = JoinHint(leftHint = Some(HintInfo(Some(NO_BROADCAST_HASH))), rightHint = None)
+            joinPlanInsert = Join(join, r, LeftOuter, Some(cond), joinHint)
+            val insertExp = EqualNullSafe(r.output.find(_.name.equals("id")).get, join.output.find(_.name.equals("id")).get)
+            joinPlanInsert = Filter(Not(insertExp), joinPlanInsert)
+            joinPlanInsert
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala:121]: this case seems useless.

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+          case _ =>
+            throw new UnsupportedOperationException("Table does not support row level operations")
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala:123]: useless code

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        }
+  }
+
+  def resolveRef[T <: NamedExpression](ref: NamedReference, plan: LogicalPlan): T = {
+    plan.resolve(ref.fieldNames.toSeq, conf.resolver) match {
+      case Some(namedExpr) =>
+        namedExpr.asInstanceOf[T]
+      case None =>
+        throw new UnsupportedOperationException("Table does not support row level operations")
+    }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala:133]: This method is duplicated with others

File: spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/expressions/AssignmentUtils.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+//    if (CharVarcharUtils.hasCharVarchar(rawKeyType)) {
+//      val newKey = key.transform {
+//        case attr: AttributeReference => CharVarcharUtils.cleanAttrMetadata(attr)
+//      }
+//      val newValue = CharVarcharUtils.stringLengthCheck(value, rawKeyType)
+//      Assignment(newKey, newValue)
+//    } else {
+//      assignment
+//    }
+//  }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/expressions/AssignmentUtils.scala:83]: remove useless code

File: spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/plans/MergeIntoArcticTable.scala [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+//
+//    newChildren match {
+//      case Seq(newTarget, newSource) =>
+//        copy(targetTable = newTarget, sourceTable = newSource, rewritePlan = None)
+//      case Seq(newTarget, newSource, newRewritePlan) =>
+//        copy(targetTable = newTarget, sourceTable = newSource, rewritePlan = Some(newRewritePlan))
+//      case _ =>
+//        throw new IllegalArgumentException("MergeIntoIcebergTable expects either two or three children")
+//    }
+//  }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/plans/MergeIntoArcticTable.scala:76]: remove useless lines

File: spark/v3.1/spark/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WritingSparkTask.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
+package org.apache.spark.sql.execution.datasources.v2
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WritingSparkTask.scala:1]: lack liences.

**Review by baiyangtx — CHANGES_REQUESTED** [review]

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+              WriteMerge(r, mergeQuery, options)
+//              mergeQuery
+            } else {
+              throw new UnsupportedOperationException("Table does not support row level operations")
+            }
+          }
+        }
+
+      case c@MergeIntoArcticTable(aliasedTable, source, cond, matchedActions, notMatchedActions, None)
+        if matchedActions.isEmpty && notMatchedActions.size == 1 && isArcticRelation(aliasedTable)=>
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteArcticMergeInto.scala:107]: It seems that case branches are not covered all situations

**Review by baiyangtx — CHANGES_REQUESTED** [review]

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/WriteMode.java [hunk 8]
```diff
@@ -22,7 +22,8 @@ public enum WriteMode {
   OVERWRITE_BY_FILTER("overwrite-by-filter"),
   OVERWRITE_DYNAMIC("overwrite-dynamic"),
   APPEND("append"),
-  UPSERT("upsert");
+  UPSERT("upsert"),
```
[the remark this answers is not in the document]
Comment (hellojinsilei) [reply spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/WriteMode.java:25]: The merge row is different from the upsert row, so I consider adding a new WriteMode to build a new writer

**Review by baiyangtx — CHANGES_REQUESTED** [review]
Unit test need exception test case.
1. merge on column without primary key.
2. merge with extra column but not specfic column names.
3. merge on primary key but matched mulit rows for one key.

and the source table need to change to non-arctic table, using an in-memory temp view instead.

and the target table need add some change data.

File: site/docs/ch/spark/spark-ddl.md [hunk 9]
```diff
@@ -206,6 +206,10 @@ ALTER TABLE arctic_catalog.db.sample ALTER COLUMN nested.col AFTER other_col;
 ALTER TABLE arctic_catalog.db.sample DROP COLUMN id;
 ALTER TABLE arctic_catalog.db.sample DROP COLUMN point.z;
 ```
+### ### ALTER TABLE ... DROP PARTITION
```
Comment (baiyangtx) [inline site/docs/ch/spark/spark-ddl.md:209]: Duplicate ###

**Review by baiyangtx — CHANGES_REQUESTED** [review]

File: site/docs/ch/spark/spark-dml.md [hunk 10]
```diff
@@ -123,6 +123,14 @@ DELETE FROM arctic_catalog.db.sample AS t1
...
 ```
 
+### TRUNCATE TABLE
+
+Arctic Spark 支持 `TRUNCATE TABLE` 语法用于删除表中所有行
+
+```sql
+TRUNCATE TABLE arctic_catalog.db.sample;
+```
+
```
Comment (baiyangtx) [inline site/docs/ch/spark/spark-dml.md:133]: TRUNCATE is a DDL statement

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/CreateArcticTableAsSelect.scala [hunk 11]
```diff
@@ -0,0 +1,108 @@
+package com.netease.arctic.spark.sql.catalyst.plans
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/CreateArcticTableAsSelect.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/MergeIntoArcticTable.scala [hunk 12]
```diff
@@ -0,0 +1,41 @@
+package com.netease.arctic.spark.sql.catalyst.plans
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/MergeIntoArcticTable.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/MergeRows.scala [hunk 13]
```diff
@@ -0,0 +1,34 @@
+package com.netease.arctic.spark.sql.catalyst.plans
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/MergeRows.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/AppendInsertDataExec.scala [hunk 14]
```diff
@@ -1,11 +1,9 @@
-package org.apache.spark.sql.execution.datasources.v2
+package com.netease.arctic.spark.sql.execution
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/AppendInsertDataExec.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/WriteMerge.scala [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package com.netease.arctic.spark.sql.catalyst.plans
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/WriteMerge.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala [hunk 16]
```diff
@@ -35,6 +36,8 @@ class ArcticSparkExtensions extends (SparkSessionExtensions => Unit) {
     }
     // resolve arctic command
     extensions.injectResolutionRule { spark => ResolveArcticCommand(spark) }
+    extensions.injectResolutionRule { spark => analysis.ResolveMergeIntoTableReferences(spark) }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/ArcticSparkExtensions.scala:39]: why need this rules?

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/ResolveMergeIntoTableReferences.scala [hunk 17]
```diff
@@ -0,0 +1,221 @@
+package com.netease.arctic.spark.sql.catalyst.analysis
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/ResolveMergeIntoTableReferences.scala:1]: License

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticStrategy.scala [hunk 18]
```diff
@@ -110,6 +118,17 @@ case class ExtendedArcticStrategy(spark: SparkSession) extends Strategy with Pre
         case table =>
           throw new UnsupportedOperationException(s"Cannot overwrite by filter to non-Arctic table: $table")
       }
+
+    case MergeRows(isSourceRowPresent, isTargetRowPresent, matchedConditions, matchedOutputs, notMatchedConditions,
+    notMatchedOutputs, targetOutput, rowIdAttrs, performCardinalityCheck, unMatchedRowCheck, emitNotMatchedTargetRows,
+    output, child) =>
+      val unMatchedRowNeedCheck = java.lang.Boolean.valueOf(spark.sessionState.conf.
+        getConfString(SparkSQLProperties.CHECK_DATA_DUPLICATES_ENABLE,
+          SparkSQLProperties.CHECK_DATA_DUPLICATES_ENABLE_DEFAULT)) && notMatchedOutputs.nonEmpty && unMatchedRowCheck
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticStrategy.scala:127]: Move this part to MergeRows object creating.

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+    val mergeRows = plans.MergeRows(
+      isSourceRowPresent = IsNotNull(rowFromSourceAttr),
+      isTargetRowPresent = if (notMatchedActions.isEmpty) TrueLiteral else IsNotNull(rowFromTargetAttr),
+      matchedConditions = matchedConditions,
+      matchedOutputs = matchedOutputs,
+      notMatchedConditions = notMatchedConditions,
+      notMatchedOutputs = notMatchedOutputs,
+      // only needed if emitting unmatched target rows
+      targetOutput = Nil,
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:189]: Is this field always be Nil? Why we need this?

**Review by baiyangtx — CHANGES_REQUESTED** [review]

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/SimpleMergeRowDataWriter.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
+package com.netease.arctic.spark.writer;
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/SimpleMergeRowDataWriter.java:1]: Liences

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/SparkInternalRowCastWrapper.java [hunk 21]
```diff
@@ -45,7 +46,96 @@ public SparkInternalRowCastWrapper(InternalRow row, StructType schema, ChangeAct
...
+    List<DataType> dataTypeList = Arrays.stream(rowSchema.fields())
+        .map(StructField::dataType).collect(Collectors.toList());
+    List<Integer> colOrdinals = JavaConverters.seqAsJavaList(projectingInternalRow.colOrdinals())
+            .stream().map(o -> Integer.parseInt(o.toString())).collect(Collectors.toList());
+    List<Object> rows = new ArrayList<>();
+    for (int i = 0; i < dataTypeList.size(); i++) {
+      rows.add(row.get(colOrdinals.get(i), dataTypeList.get(i)));
+    }
+    return new GenericInternalRow(rows.toArray());
+  }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/SparkInternalRowCastWrapper.java:71]: Is this needed? need to check

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/SimpleMergeRowDataWriter.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+  }
+
+  @Override
+  public WriterCommitMessage commit() throws IOException {
+    WriteResult result = writer.complete();
+    if (!isKeyedTable) {
+      return new WriteTaskDeleteFilesCommit(result.deleteFiles(), result.dataFiles());
+    }
+    return new WriteTaskCommit(result.dataFiles());
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/SimpleMergeRowDataWriter.java:58]: Simplify

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+  private final val ROW_FROM_TARGET = "__row_from_target"
+
+  private final val ROW_FROM_SOURCE_REF = FieldReference(ROW_FROM_SOURCE)
+  private final val ROW_FROM_TARGET_REF = FieldReference(ROW_FROM_TARGET)
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case m@MergeIntoArcticTable(aliasedTable, source, cond, matchedActions, notMatchedActions, None) =>
+
+      EliminateSubqueryAliases(aliasedTable) match {
+        case r@DataSourceV2Relation(tbl, _, _, _, _) =>
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:36]: if isArcticRelation

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+    table match {
+      case arctic: ArcticSparkTable =>
+        if (arctic.table().isKeyedTable) {
+          val primarys = arctic.table().asKeyedTable().primaryKeySpec().fieldNames()
+          val condRefs = cond.references.filter(f => primarys.contains(f.name))
+          if (condRefs.isEmpty) {
+            throw new UnsupportedOperationException(s"Condition ${cond.references}. is not allowed because is not a primary key")
+          }
+        }
+    }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:57]: Move to ResolveMergeIntoArcticTableReference rule

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+                                 isKeyedTable: Boolean): WriteQueryProjections = {
+    val (frontRowProjection, backRowProjection) = if (isKeyedTable) {
+      val frontRowProjection =
+        Some(TranslateUtils.newLazyProjection(plan, targetRowAttrs, isFront = true, 0))
+      val backRowProjection =
+        TranslateUtils.newLazyProjection(source, targetRowAttrs, isFront = false, 1 + rowIdAttrs.size)
+      (frontRowProjection, backRowProjection)
+    } else {
+      val frontRowProjection =
+        Some(TranslateUtils.newLazyProjection(plan, targetRowAttrs ++ rowIdAttrs, isFront = true, 0))
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:125]: TranslateUtils.newLazyProjection  ->  ProjectionInternalRow.newProjectInternalRow

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+      isTargetRowPresent = if (notMatchedActions.isEmpty) TrueLiteral else IsNotNull(rowFromTargetAttr),
+      matchedConditions = matchedConditions,
+      matchedOutputs = matchedOutputs,
+      notMatchedConditions = notMatchedConditions,
+      notMatchedOutputs = notMatchedOutputs,
+      // only needed if emitting unmatched target rows
+      targetOutput = Nil,
+      rowIdAttrs = keyAttrs,
+      performCardinalityCheck = isCardinalityCheckNeeded(matchedActions),
+      unMatchedRowCheck = isKeyedTable(relation),
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:192]: Simplify

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 27]
```diff
@@ -0,0 +1,269 @@
...
+    WriteQueryProjections(frontRowProjection, backRowProjection)
+  }
+
+  def buildRowIdAttrs(relation: LogicalPlan): Seq[Attribute] = {
+    val attributes = relation.output.filter(r => r.name.equals("_file") || r.name.equals("_pos"))
+    attributes
+  }
+
+  // build a rewrite plan for sources that support row deltas
+  private def buildWriteDeltaPlan(
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:139]: buildRowLevelWritePlan

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala [hunk 28]
```diff
@@ -0,0 +1,269 @@
...
+      if (exprIds.contains(attr.exprId)) {
+        None
+      } else {
+        exprIds += attr.exprId
+        Some(attr)
+      }
+    }
+  }
+
+  private def deltaActionOutput(
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala:221]: rowLevelWriteOutput

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/parser/ArcticSqlExtensionsParser.scala [hunk 29]
```diff
@@ -172,10 +182,50 @@ class ArcticSqlExtensionsParser(delegate: ParserInterface) extends ParserInterfa
           throw new ParseException(Option(sqlText), e.message, position, position)
       }
     } else {
-      delegate.parsePlan(sqlText)
+      val parsedPlan = delegate.parsePlan(sqlText)
+      parsedPlan match {
+        case p =>
+          replaceMergeIntoCommands(p)
+      }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/parser/ArcticSqlExtensionsParser.scala:189]: Check explain command

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/CreateArcticTableAsSelect.scala [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+
+            case _ =>
+              None
+          }
+        }
+      }
+    }
+
+    findField(struct, fieldNames, Nil)
+  }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/CreateArcticTableAsSelect.scala:107]: Extract as an implicat class

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/WriteMerge.scala [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.netease.arctic.spark.sql.utils.WriteQueryProjections
+import org.apache.spark.sql.catalyst.analysis.NamedRelation
+import org.apache.spark.sql.catalyst.plans.logical.{Command, LogicalPlan, V2WriteCommand}
+
+case class WriteMerge(table: NamedRelation,
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/plans/WriteMerge.scala:7]: ArcticRowLevelWrite

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/MergeRowsExec.scala [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+                          child: SparkPlan) extends UnaryExecNode {
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    if (performCardinalityCheck) {
+      // request a local sort by the row ID attrs to co-locate matches for the same target row
+      Seq(rowIdAttrs.map(attr => SortOrder(attr, Ascending)))
+    } else {
+      Seq(Nil)
+    }
+  }
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/MergeRowsExec.scala:33]: if performCardinalityCheck || unMatchedRowCheck

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/WriteMergeExec.scala [hunk 33]
```diff
@@ -0,0 +1,65 @@
+package com.netease.arctic.spark.sql.execution
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/WriteMergeExec.scala:1]: Linences

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/WriteMergeExec.scala [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.datasources.v2.{BatchWriteHelper, ExtendedV2ExistingTableWriteExec, WritingSparkTask}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+/**
+ * Physical plan node to write a delta of rows to an existing table.
+ */
+case class WriteMergeExec(
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/WriteMergeExec.scala:16]: ArcticRowLevelWriteExec

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteAppendArcticTable.scala [hunk 35]
```diff
@@ -18,16 +18,17 @@
 
 package com.netease.arctic.spark.sql.optimize
 
+import com.netease.arctic.spark.sql.catalyst.plans
 import com.netease.arctic.spark.{ArcticSparkCatalog, SparkSQLProperties}
-import com.netease.arctic.spark.sql.catalyst.plans.{AppendArcticData, OverwriteArcticData, OverwriteArcticDataByExpression, ReplaceArcticData}
+import com.netease.arctic.spark.sql.catalyst.plans.{AppendArcticData, CreateArcticTableAsSelect, OverwriteArcticData, OverwriteArcticDataByExpression, ReplaceArcticData}
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/optimize/RewriteAppendArcticTable.scala:23]: useless import

**Review by baiyangtx — CHANGES_REQUESTED** [review]
Add explain test

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/SparkInternalRowCastWrapper.java (hunk 21 above)
Comment (hellojinsilei) [reply spark/v3.1/spark/src/main/java/com/netease/arctic/spark/SparkInternalRowCastWrapper.java:71]: No, it is really unnecessary, I will remove

File: spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/UnkeyedSparkBatchWrite.java [hunk 36]
```diff
@@ -179,16 +179,16 @@ public DataWriterFactory createBatchWriterFactory(PhysicalWriteInfo info) {
...
       }
 
       AppendFiles appendFiles = table.newAppend();
-      if (WriteTaskDeleteFilesCommit.dataFiles(messages).iterator().hasNext()) {
-        for (DataFile file : WriteTaskDeleteFilesCommit.dataFiles(messages)) {
+      if (WriteTaskCommit.files(messages).iterator().hasNext()) {
+        for (DataFile file : WriteTaskCommit.files(messages)) {
           appendFiles.appendFile(file);
         }
         appendFiles.commit();
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/java/com/netease/arctic/spark/writer/UnkeyedSparkBatchWrite.java:194]: Here get some problems. why use two different iceberg op to do commit?

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/ArcticExtensionUtils.scala [hunk 37]
```diff
@@ -85,4 +165,10 @@ object ArcticExtensionUtils {
     }
   }
 
+  def isKeyedTable(relation: DataSourceV2Relation): Boolean = {
+    relation.table match {
+      case arctic: ArcticSparkTable =>
+        arctic.table().isKeyedTable
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/ArcticExtensionUtils.scala:171]: lack default match case

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/MergeRowsExec.scala [hunk 38]
```diff
@@ -147,7 +146,7 @@ case class MergeRowsExec(
       }
     }
 
-    val processFunc: InternalRow => InternalRow = if (performCardinalityCheck) {
+    val processFunc: InternalRow => InternalRow = if (performCardinalityCheck || unMatchedRowCheck) {
       processRowWithCardinalityCheck
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/MergeRowsExec.scala:150]: consider changing the method name if you are doing check for both matched key and non-matched key.

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/hive/TestKeyedTableDml.java [hunk 39]
```diff
@@ -328,4 +328,17 @@ public void testDuplicateEnabled() {
...
+  public void testExplain() {
+    sql("explain insert into {0}.{1} values (1, ''aaa'', ''aaaa'')", database, notUpsertTable);
+    sql("explain delete from {0}.{1} where id = 3", database, notUpsertTable);
+    sql("explain update {0}.{1} set name = ''aaa'' where id = 3", database, notUpsertTable);
+    rows = sql("select id, name from {0}.{1} order by id", database, notUpsertTable);
+
+    Assert.assertEquals(3, rows.size());
+    Assert.assertEquals(1, rows.get(0)[0]);
+    Assert.assertEquals(2, rows.get(1)[0]);
+    Assert.assertEquals(3, rows.get(2)[0]);
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/hive/TestKeyedTableDml.java:342]: I think the explain unit case should fetch result of explain sql, and check it is the arctic physical plan

**Review by baiyangtx — CHANGES_REQUESTED** [review]
@hellojinsilei CC

**Review by baiyangtx — APPROVED** [review]

**Pull request merged**